alf/algorithms/lagrangian_reward_weight_algorithm.py

```diff
@@ -0,0 +1,132 @@
+                its init value is always used.
+            optimizer (optimizer): optimizer for learning the reward weights.
+            init_weights (float|list[float]): the initial reward weights.
+            debug_summaries (bool):
+            name (str):
+        """
+        super(LagrangianRewardWeightAlgorithm, self).__init__(
+            debug_summaries=debug_summaries, name=name)
+
+        self._reward_spec = reward_spec
+        self._reward_thresholds = reward_thresholds
+
+        assert reward_spec.numel > 1, (
+            "Only multi-dim reward needs this algorithm!")
+        assert len(reward_thresholds) == reward_spec.numel, (
+            "Mismatch between len(reward_weights)=%s and reward_dim=%s" %
+            (len(reward_thresholds), reward_spec.numel))
+
+        self._reward_training_mask = torch.tensor(
+            [t is not None for t in reward_thresholds], dtype=torch.float32)
+        self._reward_thresholds = torch.tensor(
+            [0. if t is None else t for t in reward_thresholds])
+
+        lambda_init = torch.tensor(init_weights)
+        if lambda_init.ndim == 0:
+            lambda_init = tensor_utils.tensor_extend_new_dim(
+                lambda_init, 0, reward_spec.numel)
+        assert torch.all(
+            lambda_init >= 0.), "Initial weights must be non-negative!"
+
+        # convert to softplus space
+        self._lambdas = nn.Parameter(self._inv_softplus(lambda_init))
+        self._optimizer = optimizer
+        self._optimizer.add_param_group({'params': self._lambdas})
+
+    def _inv_softplus(self, tensor):
+        return torch.where(tensor > 20., tensor, tensor.expm1().log())
+
+    @property
+    def reward_weights(self):
+        """Return the reward weights."""
+        return F.softplus(self._lambdas)
+
+    def _trainable_attributes_to_ignore(self):
+        return ["_lambdas"]
+
+    def rollout_step(self, time_step, state):
+        return AlgStep(
+            info=LagInfo(rollout_reward=time_step.untransformed.reward))
+
+    def after_train_iter(self, experience, train_info: LagInfo):
+        """Retrieve *untransformed* rollout rewards from ``train_info``
+        and train lambdas.
+        """
+        # [T, B, reward_dim]
+        loss = ((train_info.rollout_reward - self._reward_thresholds).detach()
+                * (self.reward_weights * self._reward_training_mask))
+        loss = loss.sum(dim=-1).mean()
+
+        self._optimizer.zero_grad()
+        loss.backward()
+        self._optimizer.step()
+
+        if self._debug_summaries:
+            with alf.summary.scope(self._name):
+                alf.summary.scalar("cost", loss)
+                for i in range(len(self._reward_thresholds)):
+                    alf.summary.scalar("reward_threshold/%d" % i,
+                                       self._reward_thresholds[i])
+                    alf.summary.scalar("lambda/%d" % i, self.reward_weights[i])
```

Comment:
reward_weights here will be calculated multiple time by the property function. It's better to calculate reward_weights once it is updated at line 124.

Reply:
changed